Datum transform handling rework, pt 1

nyalldawson opened the pull request:
This PR implements part 1 of reworked datum transform handling, as discussed at Nodebo this year and at length in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/100

This PR covers a subset of the desired functionality discussed in QEP 100. It sets in place a flexible framework which can be extended in future with the other desired features discussed in the QEP (specifically, temporal handling and selection of datum transforms based on geographic extents).

In this PR:

- A new QgsCoordinateTransformContext class is implemented, for storing and managing the logic
relating to selection of datum transforms to use when transforming coordinates. This class is implicitly shared and thread safe.
- QgsCoordinateTransformContext has a mapping from source CRS to datum transform (mSourceDatumTransforms).
- QgsCoordinateTransformContext has a mapping from destination CRS to datum transform (mDestDatumTransforms).
- Additionally, it also stores a mapping of unique source/destination CRS pairs to the corresponding
datum transforms to be utilised for this source/destination combination (mSourceDestDatumTransforms).
- When constructing a QgsCoordinateTransform, if the required source to destination combination
exists in the context's mSourceDestDatumTransforms, we use that - if not, we check
mSourceDatumTransforms and mDestDatumTransforms. I.e. a matching pair of
source/destination CRS pair will take precedence, falling back to the individual transforms set for
the source and destination separately

In future:
- When more details about the dynamic coordinate systems and data formats are available,
support for dynamic temporal based transforms can be added to QgsCoordinateTransformContext - e.g. a "transform datetime" member for specifying the desired datetime which the transform should target
- QEP #100 identified the requirement that different datum transforms can be required for a single 
source and destination CRS pair, varying only in the geographic extent which each datum transform
applies to. Accordingly, in future it would be possible to add extent (QgsRectangle) or arbitrary
polygon geometry (QgsGeometry) based transform selection to QgsCoordinateTransformContext and QgsCoordinateTransform. In this case, the input coordinates would be tested against the corresponding extents stored within QgsCoordinateTransformContext and the appropriate datum transform selected based on these extents.

In this PR:
- QgsProject has a new QgsCoordinateTransformContext member, with public const/non-const
reference getters. Accordingly, datum transform settings are per-project settings, and are saved and restored along with a project. 

QgsCoordinateTransform

QgsCoordinateTransform has been upgraded to take advantage of the new QgsCoordinateTransformContext class. While ideally QgsCoordinateTransform should require all users to specify a valid QgsCoordinateTransformContext before allowing any reprojection of coordinates, it was determined that this approach is not feasible for the QGIS 3.0 release timeline. Accordingly, a “soft” approach has been taken, which allows the context to be specified but does not require it.

If the context is NOT specified by a use of QgsCoordinateTransform, then datum handling will not be
performed during coordinate reprojection for that code function. This allows us to introduce the new classes and logic, and over the QGIS 3.x lifecycle slowly upgrade the remaining components of the QGIS codebase to correctly utilise the transform context handling. 

Since upgrading PyQGIS use of QgsCoordinateTransform is straightforward (it just requires addition of a new QgsProject.instance() argument at the end of the QgsCoordinateTransform constructor), the non-context aware constructor has been removed from the Python bindings. This FORCES all PyQGIS scripts and plugins to be fully context-aware, and accordingly automatically gain correct handling for datum transformation.

I have replaced all "easy" uses of QgsCoordinateTransform in master. This includes those in app/gui (via use of QgsProject::instance() ), and those in core where the correct project instance is available. The others I have hidden the deprecation warnings for, and added qWarning debug messages to allow us to easily identify remaining parts of the code base which are not datum-transform aware.

Future work:, not covered in this part:
- Upgrading the existing datum transform behaviour from QGIS 2.x (specifically, the
transform handling which is applied during map renders, clipboard handling, and vector layer exports).
- Creation of a UI for configuring datum transformations within project properties (to be managed by @3nids)
- Creation of a UI within the global QGIS options for setting up “default” datum transformations which
new projects should inherit, or development of the code for inheriting these global settings when
creating a new project

nyalldawson committed:
Start on QgsCoordinateTransformContext

Stores settings related to the correct datum transforms to use
when performing a coordinate transform.

nyalldawson committed:
Allow removing existing transforms from context

nyalldawson committed:
Doxygen

nyalldawson committed:
QgsCoordinateTransformContext is implicitly shared

nyalldawson committed:
QgsCoordinateTransformContext must be thread safe

nyalldawson committed:
Implement logic for calculating datum transforms to use

nyalldawson committed:
Add support for serializing transform contexts

nyalldawson committed:
Add a QgsCoordinateTransformContext member to QgsProject

nyalldawson committed:
Remove unused QgsCoordinateTransform::readXml/writeXml methods

These methods are not used in master and are of questionable
value. Better to serialise the source and dest crs separately
and create the transform when required

nyalldawson committed:
Add QgsCoordinateTransformContext arguments to QgsCoordinateTransform constructor

And automatically retrieve correct datum transforms for the
transform, based on the information in the context.

Also add a convenience constructor which takes a QgsProject
instance instead of a QgsCoordinateTransformContext and which
automatically retrieves the project's transform context and uses
that. This is designed to make it easy to upgrade existing
Python code (by adding just QgsProject.instance() to the transform
constructors).

nyalldawson committed:
Remove context-unaware QgsCoordinateTransform constructors from Python bindings

This forces Python code and plugins to become datum transform
aware, and given that upgrading python code is easy (just
add QgsProject.instance() as a new argument to the constructor)
it's relatively painless to force this on PyQGIS users.

Also fix upgrade the easy QgsCoordinateTransform c++ constructors
where the project is available, or where using QgsProject::instance()
is safe to do.

For others, just avoid the deprecated warnings until we can
get access to the correct project instance where the transform
is being constructed.

nyalldawson committed:
In debug builds, throw a warning when transforming coordinates
and no coordinate transform context has been set.

m-kuhn commented on src/core/qgscoordinatetransformcontext.h:
```diff
@@ -0,0 +1,203 @@
+
+/***************************************************************************
+ * This class is considered CRITICAL and any change MUST be accompanied with
+ * full unit tests in testqgsfeature.cpp.
```
You probably meant `testqgscoordinatetransformcontext.cpp`?

m-kuhn commented on src/core/qgscrscache.cpp:
```diff
@@ -44,7 +44,9 @@ QgsCoordinateTransform QgsCoordinateTransformCache::transform( const QString &sr
   //not found, insert new value
   QgsCoordinateReferenceSystem srcCrs = QgsCoordinateReferenceSystem::fromOgcWmsCrs( srcAuthId );
   QgsCoordinateReferenceSystem destCrs = QgsCoordinateReferenceSystem::fromOgcWmsCrs( destAuthId );
+  Q_NOWARN_DEPRECATED_PUSH
```
Is there a reason for keeping it this way?

nyalldawson committed:
Add explicit setter for project's transform context

And add transformContextChanged signal, so that we can detect
when the context is changed.

nyalldawson committed:
Move warning to better location

nyalldawson committed:
Add transform context to QgsMapSettings and QgsRenderContext

And also throw warnings when the context isn't set for these objects

nyalldawson committed:
Tweak transform logic (only remove src/dest transforms when both transforms are -1)

nyalldawson committed:
Remove QgsDatumTransformStore

and port usage over to QgsCoordinateTransformContext instead

nyalldawson committed:
Prepare for removal of QgsCoordinateTransformCache

nyalldawson committed:
Move caching to QgsCoordinateTransform

and remove no longer required QgsCoordinateTransformCache singleton

nyalldawson committed:
Fix build

nyalldawson commented:
@3nids I believe this is complete now - everything but the GUI component should be done.

nyalldawson committed:
Improve caching of transforms

nyalldawson committed:
Fix failing test

nyalldawson committed:
Fix handling of transform-not-required in QgsCoordinateTransformContext

3nids commented:
I won't have time to review and do the gui part before the 20th. Is it OK for you to merge like this?

nyalldawson commented:
I'm not comfortable merging this without some form of review. @wonder-sk @m-kuhn ?

3nids committed:
started datum UI refactoring

3nids committed:
modification to QgsDatumTransformationDialog to allow interaction with CRSs

3nids committed:
adding datum transform works

3nids committed:
added some options

3nids committed:
saving datums transform in projet, removing in dialog ok

3nids committed:
ifdef 0 single source/destination transforms

3nids committed:
added method in QgisApp to ask for datum

3nids committed:
replace datum in qgsoptionsbase after rebase

3nids committed:
todos for layout after rebase (some more)

nyalldawson committed:
Fix ui build warnings

nyalldawson committed:
Update layout map to use transform context

nyalldawson committed:
Fix some transforms missing transform context

nyalldawson committed:
Update unit tests

nyalldawson committed:
Add method to determine whether transform context has a valid transform
for a specific src/dest CRS pair

3nids committed:
add missing refresh

3nids committed:
fix destination transform not showing

3nids committed:
datum transforms have reversible source/dest

3nids committed:
fix context was not correctly set on project opening

3nids committed:
use QgsProjectionSelectionWidget rather than buttons

3nids committed:
allow editing transform in project properties

3nids committed:
define default transforms in global options

adds QgsCoordinateTransformContext::readSettings()/writeSettings to read/write from settings

3nids committed:
create a QgsDatumTransformTableWidget to avoid code duplication

this contains the table view and the 3 buttons (add/remove/edit)

3nids committed:
apply default transform to newly created project

3nids committed:
fix indentation

3nids commented:
This now integrates the UI refactoring as well.
Default datum transforms can be defined in global options.
These are applied to any new project created and can be edited in the project properties.

@nyalldawson 
* I left the `#if 0` blocks in the code with the single source or destination transforms. 
* There is still a todo left in `QgsMapCanvas::layerCrsChange()`
* Do you think we can remove `QgsMapCanvas::updateDatumTransformEntries`?
* `QgisApp::askForDatumTransform` is only shown when copy/paste. would be good to add it on adding layer too I guess.
* It might be interesting to add messagebar items to warn the user about multiple transforms being available in other places of the code, but I'm not really sure here. I think that there's almost always several possible...so not sure it's worth the message here (since add layer and copy paste are already handled)

3nids commented:
I tested to remove the grid from proj and see what happens.
The transform is well listed in the datum transform table widget but the layer is not rendered.

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -6969,28 +6969,16 @@ void QgisApp::saveAsVectorFileGeneral( QgsVectorLayer *vlayer, bool symbologyOpt
       QgsSettings settings;
-      QList< QList< int > > dt = QgsCoordinateTransform::datumTransformations( vlayer->crs(), destCRS );
-      if ( dt.size() > 1 && settings.value( QStringLiteral( "Projections/showDatumTransformDialog" ), false ).toBool() )
+      QgsDatumTransformDialog *dlg = new QgsDatumTransformDialog( vlayer->crs(), destCRS );
```
No need for a raw pointer here - use a unique_ptr or not a pointer

3nids replied:
fixed

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -12379,6 +12372,47 @@ void QgisApp::writeDockWidgetSettings( QDockWidget *dockWidget, QDomElement &ele
+  {
+    //if several possibilities:  present dialog
+    QgsDatumTransformDialog dlg( sourceCrs, destinationCrs );
+    if ( dlg.availableTransformationCount() > 1 )
```
One thing I'm not sure of with this approach - wouldn't this mean that if the user only has one valid transform, they don't get the dialog, and according the project's context is never populated with that one transform?

3nids replied:
that's correct. Do you see any situation where it could be problematic?

nyalldawson commented on src/app/qgsdatumtransformtablewidget.cpp:
```diff
@@ -0,0 +1,252 @@
+    int destinationTransform = -1;
+    for ( QModelIndexList::const_iterator it = selectedIndexes.constBegin(); it != selectedIndexes.constEnd(); it ++ )
+    {
+      if ( it->column() == QgsDatumTransformTableModel::SourceCrsColumn )
```
a switch would be nicer here

3nids replied:
fixed

nyalldawson commented on src/app/qgsdatumtransformtablewidget.cpp:
```diff
@@ -0,0 +1,252 @@
+    if ( sourceCrs.isValid() && destinationCrs.isValid() &&
+         ( sourceTransform != -1 || destinationTransform != -1 ) )
+    {
+      QgsDatumTransformDialog *dlg = new QgsDatumTransformDialog( sourceCrs, destinationCrs, qMakePair( sourceTransform, destinationTransform ) );
```
dlg is leaked here

3nids replied:
fixed

nyalldawson commented on src/app/qgsdatumtransformtablewidget.h:
```diff
@@ -0,0 +1,104 @@
+
+    QgsDatumTransformTableModel( QObject *parent = nullptr );
+
+    void setTransformContext( QgsCoordinateTransformContext &context );
```
const ref?

3nids replied:
fixed

nyalldawson commented on src/app/qgsdatumtransformtablewidget.h:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * remove the transformation at given indexes
+     */
+    void removeTransform( QModelIndexList indexes );
```
const ref

3nids replied:
fixed

nyalldawson commented on src/app/qgsdatumtransformtablewidget.h:
```diff
@@ -0,0 +1,104 @@
+    explicit QgsDatumTransformTableWidget( QWidget *parent = 0 );
+    ~QgsDatumTransformTableWidget();
+
+    void setTransformContext( QgsCoordinateTransformContext &context )
```
const QgsCoordinateTransformContext& ?

3nids replied:
fixed

3nids committed:
address missing const and memory leaks

kudos @nyalldawson

3nids committed:
use alternate row colors and a slightly reduced font size in datum table

3nids commented on src/core/qgscoordinatetransformcontext.cpp:
```diff
@@ -257,7 +266,92 @@ void QgsCoordinateTransformContext::writeXml( QDomElement &element, QDomDocument
+  d->mDestDatumTransforms.clear();
+#endif
+
+  QgsSettings *settings = new QgsSettings();
```
fixed

3nids committed:
missing const ref

3nids commented on src/gui/qgsdatumtransformdialog.cpp:
```diff
@@ -258,36 +258,18 @@ void QgsDatumTransformDialog::mDatumTransformTreeWidget_currentItemChanged( QTre
 }
 
-void QgsDatumTransformDialog::setDestinationCrs()
+void QgsDatumTransformDialog::setDestinationCrs( QgsCoordinateReferenceSystem destinationCrs )
```
fixed

3nids commented on src/gui/qgsdatumtransformdialog.cpp:
```diff
@@ -258,36 +258,18 @@ void QgsDatumTransformDialog::mDatumTransformTreeWidget_currentItemChanged( QTre
 }
 
-void QgsDatumTransformDialog::setSourceCrs()
+void QgsDatumTransformDialog::setSourceCrs( QgsCoordinateReferenceSystem sourceCrs )
```
fixed

nyalldawson commented:
> There is still a todo left in QgsMapCanvas::layerCrsChange()
> Do you think we can remove QgsMapCanvas::updateDatumTransformEntries?

I think updateDatumTransformEntries() should be bumped up to QgisApp (so it can call askForDatumTransform), and the same with the detection of project layerCrsChanges (also call askForDatumTransform).

> QgisApp::askForDatumTransform is only shown when copy/paste. would be good to add it on adding layer too I guess.

Agreed - I'd insert this into the various QgisApp addVectorLayer/addRasterLayer methods, so that it's only triggered as a result of users adding layers (vs plugins/scripts opening layers)

> It might be interesting to add messagebar items to warn the user about multiple transforms being available in other places of the code, but I'm not really sure here. I think that there's almost always several possible...so not sure it's worth the message here (since add layer and copy paste are already handled)

Let's leave this off for now and address if there's a concrete need for it shown.

> I tested to remove the grid from proj and see what happens. The transform is well listed in the datum transform table widget but the layer is not rendered.
 
Did you rerun the crs sync after doing this? If it's being listed by QgsCoordinateTransform::datumTransformations then I suspect not...


I've got some more commits inbound, which clean up more of the datum transform API (switching the structs instead of QPairs, extra docs, etc).

nyalldawson committed:
Nicer API for datum transforms

Instead of using QPairs of ints, use more descriptive structs, also
rename a lot of datum related methods for clarity and add docs

nyalldawson committed:
Add some more datum related unit tests

nyalldawson commented:
Can we get a volunteer to review this PR (specifically core changes)? @3nids and I have discussed it at length, but I think something so fundamental as this needs another pair of eyes.

3nids committed:
address remaining memory leaks

3nids committed:
ask for datum transform update on layer or project crs change

in the case of project crs change, if several transforms are available, pop up a message rather than showing multiple dialogs

m-kuhn commented on src/app/qgisapp.cpp:
```diff
@@ -9012,6 +9005,45 @@ void QgisApp::onFocusChanged( QWidget *oldWidget, QWidget *newWidget )
+  // handle datum transforms
+  QList<QgsCoordinateReferenceSystem> transformsToAskFor = QList<QgsCoordinateReferenceSystem>();
+  QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
+  for ( QMap<QString, QgsMapLayer *>::iterator it = layers.begin(); it != layers.end(); ++it )
```
const_iterator?

m-kuhn commented on src/app/qgisapp.h:
```diff
@@ -1571,6 +1571,9 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
 
     void onFocusChanged( QWidget *oldWidget, QWidget *newWidget );
 
+    //! handles project crs changes
```
Use imperative ;)=

/**
 * Handle project crs changes
 */

m-kuhn commented on src/app/qgisapp.cpp:
```diff
@@ -12402,6 +12428,47 @@ void QgisApp::writeDockWidgetSettings( QDockWidget *dockWidget, QDomElement &ele
   elem.setAttribute( QStringLiteral( "area" ), dockWidgetArea( dockWidget ) );
 }
 
+bool QgisApp::askForDatumTransform( QgsCoordinateReferenceSystem sourceCrs, QgsCoordinateReferenceSystem destinationCrs )
```
Is it safe to call this out of the main thread?

I think it would be good to add a `Q_ASSERT` on top to get an immediate feedback for unsafe handling in dev environments.

Another thought is to add a non-modal version (especially for out of main-thread events), where right now a default config is used and the user is informed about a missing configuration for these CRS'es with a button/link to fix this missing definition later on.

3nids replied:
added Q_ASSERT

there is a similar case handled for project CRS changed which triggers `QgisApp::projectCrsChanged`

m-kuhn commented on src/app/qgisapp.h:
```diff
@@ -653,6 +653,14 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
+      * \returns true if a datum transform has been specifically chosen by user or only one is available.
+      * \since 3.0
+      */
+    bool askForDatumTransform( QgsCoordinateReferenceSystem sourceCrs, QgsCoordinateReferenceSystem destinationCrs );
```
Const reference?

    const QgsCoordinateReferenceSystem &

3nids replied:
fixed

3nids committed:
use const_iterator

m-kuhn commented on src/core/qgscoordinatetransformcontext.h:
```diff
@@ -0,0 +1,270 @@
+    void clear();
+
+
+#if 0
```
Is it planned to use this in the future? I.e. Is there any value in keeping this code?

m-kuhn commented on src/core/qgscoordinatetransformcontext.h:
```diff
@@ -0,0 +1,270 @@
+     * Reads the context's state from a DOM \a element.
+     * \see writeXml()
+     */
+    void readXml( const QDomElement &element, const QDomDocument &document, const QgsReadWriteContext &context );
```
Drop the `document` parameter and use `element.ownerDocument()` instead.

m-kuhn commented on src/core/qgscoordinatetransformcontext.h:
```diff
@@ -0,0 +1,270 @@
+     * Writes the context's state to a DOM \a element.
+     * \see readXml()
+     */
+    void writeXml( QDomElement &element, QDomDocument &document, const QgsReadWriteContext &context ) const;
```
Drop the `document` parameter and use `element.ownerDocument()` instead.

m-kuhn commented on src/core/qgscoordinatetransformcontext_p.h:
```diff
@@ -0,0 +1,87 @@
+     */
+    QMap< QPair< QString, QString >, QgsCoordinateTransform::TransformPair > mSourceDestDatumTransforms;
+
+#if 0
```
Leftover or left on purpose?

m-kuhn commented on src/app/qgsdatumtransformtablewidget.cpp:
```diff
@@ -0,0 +1,265 @@
+{
+  Q_UNUSED( parent );
+  return mTransformContext.sourceDestinationDatumTransforms().count()
+#ifdef singlesourcedest
```
How is this controlled, why is this used?

3nids replied:
all the code commented out is left from first implementation where we thought we would need several containers for transforms (source only, destination only, source & destination).
After some thoughts, I think source & destination is the only required container:
- transforms are reversible: it makes no sense to distinguish source and destination
- I don't think there is a practical use case for source or destination only transform (i.e. going to WGS84 only). 
But I'm not 100% sure for the latter, I decided to keep the code commented in case the reviewer -- you ;) -- thinks it's better to bring that back in.

m-kuhn commented:
I quickly skimmed through the code and left some comments and questions, an in-depth review is out of scope right now.

One thing I'm not completely sure, as far as I can see, configuration is always done via QgisInterface / QgisApp when adding a layer. If a configuration is missing, how is a user informed about this? And how is this supposed to happen in standalone apps?

I was expecting to have some signal (not sure what would emit that though) or callback when a missing configuration is encountered. In there some handwritten code should be able to directly fix things or print a warning for fixing things down the road.
E.g. QGIS server should be able to aggregate missing information somewhere to display on an admin interface. Or a standalone app open a different dialog when a configuration is missing.
Did I just miss this, or what is the plan?

nyalldawson commented:
@3nids 

Thought of another issue we may need to address before merge. At the moment we store datum transform IDs in the project. But I'm having trouble determining if these IDs are guaranteed to be consistent across different installations. E.g. a QGIS install on one desktop vs another on a different desktop, with different grids installed. If they're not, we'd need to store the datum string instead of the ID in the project.

3nids commented:
@m-kuhn as we discussed in Nodebo, this is mainly reserved to advanced user. Dialog will never pop if one doesn't tick the "ask for datum transforms" in global options. I am not sure showing messages will improve the situation. @nyalldawson ?

@nyalldawson indeed you're probably right. For the my swiss grid, I get id 10001, which I guess will be used for any other user-specific grid. Would you propose to use the description as string?

3nids committed:
missing const ref, remove extra argument

3nids committed:
fix transform context test

nyalldawson commented:
> @m-kuhn as we discussed in Nodebo, this is mainly reserved to advanced user. Dialog will never pop if one doesn't tick the "ask for datum transforms" in global options. I am not sure showing messages will improve the situation. @nyalldawson ?

I've no strong opinion here, but think we should address this only when there's demonstrated need for it.

> @nyalldawson indeed you're probably right. For the my swiss grid, I get id 10001, which I guess will be used for any other user-specific grid. Would you propose to use the description as string?

I think so. I'll investigate.

m-kuhn commented:
> I've no strong opinion here, but think we should address this only when there's demonstrated need for it.

The question was mainly if the api is prepared for this.

nyalldawson committed:
Fix warning

3nids commented:
not to self: when adding several layers (21781) on an empty project (4326), the dialog is shown several times, should not.
also: shall the empty project be switched to first added layer CRS?

nyalldawson committed:
Sipify

nyalldawson committed:
Update unit test

nyalldawson committed:
Add method to retrieve datum transform ID from a proj string

nyalldawson committed:
Use proj strings to serialize datum transforms IDs

Since datum transforms IDs are local only, and can even change
when underlying library versions are updated, we can't use them
to safely store and retrieve datum transforms.

Instead we use the proj strings and recover the local session
transform ID from that.

nyalldawson committed:
Show a warning when loading a project with a transform not available locally

nyalldawson commented:
@3nids latest comments addressed... that's my todo-list cleared.

nyalldawson committed:
Add a unit test for missing datum transforms

nyalldawson commented:
I'm thinking we should merge this now, and get more feedback from wider testing. Do you concur @3nids?

3nids commented:
Great job, that looks perfect to me :)
We can still change some core mechanism later, but it would be nice that the UX gets tested as soon as possible.

nyalldawson commented:
Alright... let's see how this goes. Fingers crossed!

m-kuhn commented:
Good job everyone !!! Pure excitement !!!

nirvn commented:
OK, I've just had my first encounter with this intimidating fellow upon project loading:
![screenshot from 2017-12-18 12-08-51](https://user-images.githubusercontent.com/1728657/34090823-4ec3e4fe-e3ec-11e7-823d-6b83ee819d7b.png)

If it intimidated me, I'm assuming it'll scare the average Joe 😉 I do not have [ ] ask for datum transformation if several... checked in my settings. Is there an issue with existing projects load triggering this dialog when it should just silently pick the best option?

Also, re UX, some ideas to improve things:
- src transform / dst transform columns should expand to fit content, ATM it's just impossible to use without resizing columns
- a default (best option) item should be selected instead of none
- possibly an help label below the list to explain this dialog a bit?
- column headers should probably use "source" instead of "src" (i.e. source transform instead of src transform)

nirvn commented:
Oh also, the "Source CRS" and "Destination CRS" labels should expand horizontally, the size should be fixed to allow for the projection selection widget to expand.

NathanW2 commented:
Does the bottom panel need to be a table like that? I think two text boxes
might look better there.

On Mon, Dec 18, 2017 at 3:15 PM, Mathieu Pellerin <notifications@github.com>
wrote:

> Oh also, the "Source CRS" and "Destination CRS" labels should expand
> horizontally, the size should be fixed to allow for the projection
> selection widget to expand.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/5535#issuecomment-352328224>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3AhUKT7ot6GzV7pJk5Ud9QmFa4Nvks5tBfUHgaJpZM4QSWzA>
> .
>

nirvn commented:
@nyalldawson , @3nids , one more issue: if I hit the [ cancel ] button in the above mentioned intimidating dialog, rendering of layers become *extremely* slow as the console is being flooded by the following warning: `WARNING	No QgsCoordinateTransformContext context set for transform`

nirvn commented:
@nyalldawson , @3nids , I think the dialog is triggered open loading a project when this bit is in the project file:

```
    <layer_coordinate_transform_info>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="OSM_PhnomPenh_Roads_2de9f21a_6ab9_42a4_988a_d910824fde9d"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:4326" destAuthId="EPSG:4326" layerid="Transitions_100E_20N_ad9960fd_fe00_4cd4_abbe_0f24f5d94a09"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:4326" destAuthId="EPSG:4326" layerid="Occurrence_100E_20N_8ec65e27_c0f5_4380_b248_43415bd6f80d"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="Khm_Admbnda_Adm1_Gov_665dd60c_a292_4af3_94d0_ea78d67745a4"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3857" destAuthId="EPSG:4326" layerid="Osm_e6b9c0d3_e024_42b3_aa77_82e1648017a3"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="labels_4883c325_dad7_433d_ae68_75130b948b32"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="Cambodia_NEC_2012_Geo_Indian1960_48N_1e9fc321_f865_4ddd_8638_c3e7b0469b55"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="CKHET_b64b66d2_fda6_4898_ac61_9f4e128b01e5"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:4326" destAuthId="EPSG:4326" layerid="Transitions_100E_20N_a1708ca6_1dd6_4b32_be2c_af8dc04fc004"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3148" destAuthId="EPSG:4326" layerid="Cambodia__Boundary_In_Indochina_d975d529_5e9e_4a58_9060_eaf9e4f6e410"/>
      <layer_coordinate_transform srcDatumTransform="-1" destDatumTransform="-1" srcAuthId="EPSG:3857" destAuthId="EPSG:4326" layerid="Osm_8ef47baa_5dac_4dd0_8947_3799ea39f73c"/>
    </layer_coordinate_transform_info>
```

Would this trigger the dialog upon loading a 2.X-saved project? If so, we should probably not :)

3nids commented:
@nirvn thanks for reporting, looking into this.

3nids commented:
https://github.com/qgis/QGIS/pull/5905 addresses some of these.
I'll let @nyalldawson handle the log flooding.

I don't think we can easily define a "best" option, or that's beyond my competences (and the scope of this work).
I would be in favor of writing a short text to explain, but maybe in the main options (and project options) where you call the dialog from. Thoughts?

3nids commented:
@nirvn the fixes are merged, let us know if it improves the situation. Thanks again.
